lib/pengine/utils.c

```diff
@@ -2392,13 +2392,37 @@ pe__clear_resource_history(pe_resource_t *rsc, pe_node_t *node,
 bool
 pe__rsc_running_on_any_node_in_list(pe_resource_t *rsc, GList *node_list)
 {
-    for (GList *ele = rsc->running_on; ele; ele = ele->next) {
-        pe_node_t *node = (pe_node_t *) ele->data;
-        if (pcmk__str_in_list(node_list, node->details->uname)) {
+    GList *hosts = NULL;
+
+    if (rsc == NULL) {
+        return false;
+    }
+
+    rsc->fns->location(rsc, &hosts, TRUE);
+
+    for (GList *hIter = hosts; node_list != NULL && hIter != NULL; hIter = hIter->next) {
+        pe_node_t *node = (pe_node_t *) hIter->data;
+
+        if (pcmk__str_in_list(node_list, node->details->uname) ||
+            pcmk__str_in_list(node_list, node->details->id)) {
+            crm_trace("Resource %s is running on %s", rsc->id, node->details->uname);
+            g_list_free(hosts);
             return true;
         }
     }
 
+    if (hosts != NULL) {
+        /* We could have been given more than one node here, which makes
+         * logging all the nodes it's not running on more difficult.  Maybe
+         * just logging the first will cover most cases.
+         */
+        pe_node_t *node = (pe_node_t *) hosts->data;
+        crm_trace("Resource %s is not running on: %s", rsc->id, node->details->uname);
+    } else if (hosts == NULL && node_list == NULL) {
```

Comment:
Fixed, I think.

Incidentally, this is a fairly complicated function that would be great to have some unit tests for.  The difficulty there is that it uses a pe_resource_t object and the list of pe_node_t objects.  So we'd have to come up with a way to mock up objects.  There's no way to do that in glib that I can tell, but there's the cmocka library which doesn't have any other dependencies.  That might be something to look into later.

Reply:
Actually, there's an additional problem here.  Returning true if node_list is NULL and the resource is running anywhere breaks the test for `crm_mon -1 --node=blah`.  In that case, we end up displaying all the nodes instead of none of them.

Reply:
...which can probably be taken care of entirely within pcmk__rsc_filtered_by_node.